### PR TITLE
Operators string-represent themselves.

### PR DESCRIPTION
**Pull request by smite**
**Context:**

Both `OperationRecorder` and `CircuitGraph` have methods for printing the contents of the circuit in string format, with almost identical code.

**Description of the Change:**

* The string representation of `Operators` is moved to `Operator.__repr__`

**Benefits:**

* `Tensors` are represented properly. 
* Code readability, clarity.

**Future work:**

At some point it might make sense to make a variant of `OperationRecorder` a base class for `BaseQNode`?

### CODE REVIEW

File: pennylane/operation.py [hunk 1]
```diff
@@ -164,6 +164,12 @@ class ObservableReturnTypes(Enum):
 """Enum: An enumeration which represents returning probabilities
 of all computational basis states."""
 
+_measurement_type_map = {
+    Expectation: "expval",
+    Variance: "var",
+    Sample: "sample",
+}
```
Comment (josh146) [inline pennylane/operation.py]: I'm wondering if it's possible to define the `__repr__` for an enum? That way it can be done in the enum directly.

File: pennylane/operation.py [hunk 2]
```diff
@@ -331,9 +337,19 @@ def __init__(self, *params, wires=None, do_queue=True):
...
     def __str__(self):
-        """Print the operator name and some information."""
+        """Operator name and some information."""
         return  "{}: {} params, wires {}".format(self.name, len(self.params), self.wires)
 
+    def __repr__(self):
+        """Constructor-call-like representation."""
+        # FIXME using self.parameters here instead of self.params is dangerous, it assumes the params can be evaluated
+        # which is only true if something suitable happens to remain in VariableRef.positional_arg_values etc. after
+        # the last evaluation.
```
Comment (josh146) [inline pennylane/operation.py]: Is there a valid user pattern where this is not the case? Could add an xfail test for it.

File: pennylane/operation.py [hunk 3]
```diff
@@ -726,6 +749,10 @@ def __str__(self):
         """Print the tensor product and some information."""
         return 'Tensor product {}: {} params, wires {}'.format([i.name for i in self.obs], len(self.params), self.wires)
 
+    def __repr__(self):
+        """Constructor-call-like representation."""
+        return "Tensor(" +", ".join([repr(o) for o in self.obs]) +")"
```
Comment (josh146) [inline pennylane/operation.py]: Do we want tensors to print as `Tensor(...)` though? The actual tensor class is quite hidden from users, and we only document using `@` to create tensors. This is partly because instantiating the tensor opens us up to some known buggy edge cases, that are not possible using `@` (such as tensors of single observables)

File: tests/qnodes/test_qnode_base.py [hunk 4]
```diff
@@ -173,8 +173,8 @@ def circuit(x):
             Observables
             ===========
             expval(PauliX(wires=[0]))
-            var(Hermitian([[0 1]
-             [1 0]], wires=[1]))"""
+            var(Hermitian(array([[0, 1],
+                   [1, 0]]), wires=[1]))"""
```
Comment (josh146) [inline tests/qnodes/test_qnode_base.py]: :+1:

File: pennylane/operation.py (hunk 1 above)
Comment (smite) [reply pennylane/operation.py]: It is, do you think it'd look nicer?
You could even get rid of the dict if you changed the names of the enum members to expval, var, sample etc.

File: pennylane/operation.py (hunk 2 above)
Comment (smite) [reply pennylane/operation.py]: It seems quite unsafe, say you have two qnodes, evaluate both, and then call `print_applied` on both. The method will try to evaluate the parameters of both qnodes, but only the second one has its parameter values stored in `VariableRef.positional_arg_values`. The first one will get the wrong values, or may even request a value with an index that's too high.

That's the inherent downside of using global variables, you cannot assume they will stay valid between function calls.

Comment (smite) [reply pennylane/operation.py]: I can add a test but we should think of a way of fixing this somehow.

File: pennylane/operation.py (hunk 1 above)
Comment (josh146) [reply pennylane/operation.py]: The capital in the name was chosen for convention (based on the python Enum docs), I worry changing it to lowercase might cause ambiguities to people reading the code. But I'm on the fence!

File: pennylane/operation.py (hunk 2 above)
Comment (josh146) [reply pennylane/operation.py]: Yep, I agree this should be fixed. This could also help fix some race conditions we are seeing with parallelization.

File: pennylane/operation.py (hunk 3 above)
Comment (smite) [reply pennylane/operation.py]: Hmm. I'd say do whatever serves the purpose best. `__repr__` often returns something that looks like a constructor call, but it's not a hard rule.

File: pennylane/operation.py (hunk 1 above)
Comment (smite) [reply pennylane/operation.py]: How about this, using the strings as the enum values?

File: pennylane/operation.py (hunk 2 above)
Comment (smite) [reply pennylane/operation.py]: One solution would be to remove the class variables `VariableRef.positional_arg_values` etc. and instead let each QNode keep their own "last evaluated parameters" instance variables. The `VariableRef` instances would have to have references to them so they can find them, but that should be no problem. This is more or less how Strawberry fields does this. What do you think?

File: pennylane/operation.py (hunk 3 above)
Comment (smite) [reply pennylane/operation.py]: Shall I change it?

File: pennylane/operation.py (hunk 2 above)
Comment (josh146) [reply pennylane/operation.py]: I think that sounds good - are there any downsides to changing to this approach? Sounds like this should be an ADR first.

File: pennylane/operation.py (hunk 3 above)
Comment (josh146) [reply pennylane/operation.py]: I'm on the fence, but lean towards the output matching how the user would construct it, so `@` :slightly_smiling_face: But it's not too much of an issue either way.

File: pennylane/operation.py (hunk 1 above)
Comment (josh146) [reply pennylane/operation.py]: Nice! I didn't know `repr` was a function :thinking:

**Comment by johannesjmeyer** [comment]
> At some point it might make sense to make a variant of `OperationRecorder` a base class for `BaseQNode`?

This is related to what I propose in the queuing ADR. There should at least be a common base class to both `OperationRecorder` and `BaseQNode` that handles the queuing and that provides the context methods

File: pennylane/operation.py (hunk 3 above)
Comment (johannesjmeyer) [reply pennylane/operation.py]: We could also define both `__repr__` and `__str__`, where `__repr__` is the constructor-like looking variant and `__str__` the user-friendly one with `@`

**Comment by josh146** [comment]
@smite, is CodeCov just being buggy? If so, worth just checking the coverage locally, and if all looks okay, I can merge it in.

File: pennylane/operation.py (hunk 2 above)
Comment (smite) [reply pennylane/operation.py]: I can make an ADR and see if I can think of downsides. Maybe one more pointer dereference per evaluation?

File: pennylane/operation.py (hunk 3 above)
Comment (smite) [reply pennylane/operation.py]: `__str__` is already defined, it's on line 747 just above, and it's even wordier.

**Comment by josh146** [comment]
@smite, is this ready to be merged?
